server/data/data-postgres/src/main/java/org/apache/james/droplists/postgres/PostgresDropList.java

```diff
@@ -0,0 +1,127 @@
+import static org.apache.james.droplists.postgres.PostgresDropListModule.PostgresDropListsTable.DENIED_ENTITY_TYPE;
+import static org.apache.james.droplists.postgres.PostgresDropListModule.PostgresDropListsTable.DROPLIST_ID;
+import static org.apache.james.droplists.postgres.PostgresDropListModule.PostgresDropListsTable.OWNER;
+import static org.apache.james.droplists.postgres.PostgresDropListModule.PostgresDropListsTable.OWNER_SCOPE;
+import static org.apache.james.droplists.postgres.PostgresDropListModule.PostgresDropListsTable.TABLE_NAME;
+
+import java.util.List;
+import java.util.UUID;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Named;
+import jakarta.mail.internet.AddressException;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.core.Domain;
+import org.apache.james.core.MailAddress;
+import org.apache.james.droplists.api.DropList;
+import org.apache.james.droplists.api.DropListEntry;
+import org.apache.james.droplists.api.OwnerScope;
+import org.jooq.Record;
+
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresDropList implements DropList {
+    private final PostgresExecutor postgresExecutor;
+
+    @Inject
+    public PostgresDropList(@Named(DEFAULT_INJECT) PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    @Override
+    public Mono<Void> add(DropListEntry entry) {
+        Preconditions.checkArgument(entry != null);
+        String specifiedOwner = entry.getOwnerScope().equals(OwnerScope.GLOBAL) ? "" : entry.getOwner();
+        return postgresExecutor.executeVoid(dslContext ->
+            Mono.from(dslContext.insertInto(TABLE_NAME, DROPLIST_ID, OWNER_SCOPE, OWNER, DENIED_ENTITY_TYPE, DENIED_ENTITY)
+                .values(UUID.randomUUID(),
+                    entry.getOwnerScope().name(),
+                    specifiedOwner,
+                    entry.getDeniedEntityType().name(),
+                    entry.getDeniedEntity())
+            )
+        );
+    }
+
+    @Override
+    public Mono<Void> remove(DropListEntry entry) {
+        Preconditions.checkArgument(entry != null);
+        return postgresExecutor.executeVoid(dsl -> Mono.from(dsl.deleteFrom(TABLE_NAME)
+            .where(OWNER_SCOPE.eq(entry.getOwnerScope().name()))
+            .and(OWNER.eq(entry.getOwner()))
+            .and(DENIED_ENTITY.eq(entry.getDeniedEntity()))));
+    }
+
+    @Override
+    public Flux<DropListEntry> list(OwnerScope ownerScope, String owner) {
+        Preconditions.checkArgument(ownerScope != null);
+        Preconditions.checkArgument(owner != null);
+        return postgresExecutor.executeRows(dsl -> Flux.from(dsl.selectFrom(TABLE_NAME)
+                .where(OWNER_SCOPE.eq(ownerScope.name()))
+                .and(OWNER.eq(owner))))
```

Comment:
We should add an index for (OWNER_SCOPE, OWNER) IMO.

Maybe another one for (OWNER_SCOPE, OWNER, DENIED_ENTITY)

Reply:
thank you for your valuable comments!